lib2/Ppmd8Tdecoder.c

```diff
@@ -0,0 +1,138 @@
+//
+// Created by miurahr on 2021/08/07.
+//
+
+#include "Ppmd8Tdecoder.h"
+
+PPMD_pthread_mutex_t mutex;
+PPMD_pthread_cond_t finished, inEmpty, notEmpty;
+
+Byte TReader(const void *p) {
+    BufferReader *bufferReader = (BufferReader *)p;
+    if (bufferReader->inBuffer->pos == bufferReader->inBuffer->size) {
+        PPMD_pthread_cond_signal(&inEmpty);
+        PPMD_pthread_cond_wait(&notEmpty, &mutex);
+    }
+    return *((const Byte *)bufferReader->inBuffer->src + bufferReader->inBuffer->pos++);
+}
+
+Bool Ppmd8T_decode_init() {
+    PPMD_pthread_mutex_init(&mutex, NULL);
+    PPMD_pthread_cond_init(&finished, NULL);
+    PPMD_pthread_cond_init(&inEmpty, NULL);
+    PPMD_pthread_cond_init(&notEmpty, NULL);
+    return True;
+}
+
+static void *
+Ppmd8T_decode_run(void *p) {
+    ppmd8_args *args = (ppmd8_args *)p;
+    PPMD_pthread_mutex_lock(&mutex);
+    CPpmd8 * cPpmd8 = args->cPpmd8;
+    BufferReader *reader = (BufferReader *) cPpmd8->Stream.In;
+    InBuffer *in = reader->inBuffer;
+    int max_length = args->max_length;
+    args->finished = False;
```

Comment:
variable args is not just arguments, but it is a structure to keep thread status in `finished` field.
It would be given better name.